### DISCUSSION TITLE
Facter and ffi with 64 bit ruby on AIX

### DISCUSSION THREAD

**Discussion by SaltyAlan** [puppetlabs/facter#2685] [Q&A]
I am running puppet 7.28 open source - trying to install version 4.5.2 facter on AIX, installed with ruby 2.7 gems. 

We have been running facter version 2 historically, which does not use ffi. 

There looks to be a major shift in the design of facter between version 2 and 4 - moving several facts from facter itself in to ffi. I am having trouble getting this facter to ffi integration to work on AIX. 

As AIX doesn't have puppet, facter or ffi packaged, I installed these using gems. ffi actually runs compiles using gcc at this point, it is not a simple install. 

After install using 64bit ruby, facter errors out with the following errors:

```
# facter
[2024-03-06 16:12:25.279458 ] ERROR Facter::Resolvers::Aix::Networking - Resolving fact interfaces, but got Could not open library 'libc.a(shr.o)':         0509-022 Cannot load module /usr/lib/libc.a(shr.o).
        0509-103   The module has an invalid magic number..
Could not open library 'libc.a(shr.o).so':      0509-022 Cannot load module .
        0509-026 System error: A file or directory in the path name does not exist..
Searched in <system library path>, /usr/lib, /usr/local/lib, /opt/local/lib at /opt/freeware/share/ruby/gems/2.7/gems/ffi-1.16.3/lib/ffi/dynamic_library.rb:65:in `load_library'
```

The invalid magic number error looks to point to a 32bit/64bit mismatch. The library it is looking for shr.o in /usr/lib/libc.a - this is the 32bit version of the library. 

Line 23 of the file ffi_helper.rb here:
https://github.com/puppetlabs/facter/blob/main/lib/facter/resolvers/aix/ffi/ffi_helper.rb
seems to be hard coded to point to the 32bit AIX libc library. 

`ffi_lib 'libc.a(shr.o)'`

The 64bit library on AIX is called shr_64.o . 

```
# ar -tv -X32 /usr/lib/libc.a shr.o
r-xr-xr-x     2/2     1029439 Feb 07 00:43 2023 shr.o
# ar -tv -X64 /usr/lib/libc.a shr_64.o
r-xr-xr-x     2/2     1150180 Feb 07 00:43 2023 shr_64.o
```

I tried changing this line to refer to shr_64.o on my install, - it still did not work, but errors changed to:

```
# facter
[2024-03-06 16:17:12.430064 ] ERROR Facter::Resolvers::Aix::Networking - Resolving fact interfaces, but got undefined method `read_string' for nil:NilClass at /opt/freeware/share/ruby/gems/2.7/gems/facter-4.5.2/lib/facter/resolvers/aix/ffi/ffi_helper.rb:104:in `read_interfaces'
```
so I expect something else needs to be done to get facter and ffi to work using 64bit ruby. 

My troubleshooting got puppet, facter and ffi to work by running puppet and facter with 32bit ruby instead of the default 64bit ruby, and reinstalling ffi with 32bit ruby so it compiled the ffi components in 32bit mode. 

I would like to know if it is expected that on AIX we run puppet and facter using 32bit ruby? The default is 64bit ruby which we have been using up until now.  

If this is unexpected behaviour and facter with ffi should be working with 64bit ruby, I would appreciate any assistance in getting this to work in 64bit mode. 

Thanks,
Alan

**Comment by joshcooper** [comment]
I would try ffi 1.15.5 as there are some issues with newer versions. Also recommend newer facter version too.

**Reply by SaltyAlan** [reply]
Thanks for the response. I have tried this with facter 4.6.1 and ffi 1.15.5 - still the same results with 64bit ruby. 

```
# gem install --no-document facter-4.6.1.gem ffi-1.15.5.gem
Successfully installed facter-4.6.1
Building native extensions. This could take a while...
Successfully installed ffi-1.15.5
2 gems installed

# facter
[2024-04-11 17:16:34.545452 ] ERROR Facter::Resolvers::Aix::Networking - Resolving fact interfaces, but got Could not open library 'libc.a(shr.o)':         0509-022 Cannot load module /usr/lib/libc.a(shr.o).
        0509-103   The module has an invalid magic number..
Could not open library 'libc.a(shr.o).so':      0509-022 Cannot load module .
        0509-026 System error: A file or directory in the path name does not exist. at /opt/freeware/share/ruby/gems/2.7/gems/ffi-1.15.5/lib/ffi/library.rb:145:in `block in ffi_lib'
...
...
```

Changing ruby to use 32bit:
```
# ls -l ruby
lrwxrwxrwx    1 root     system            7 Apr 10 21:33 ruby -> ruby_64
# rm ruby
# ln -s ruby_32 ruby
# ls -l ruby
lrwxrwxrwx    1 root     system            7 Apr 11 17:21 ruby -> ruby_32
```

Reinstalling facter and ffi:

```
# gem install --no-document facter-4.6.1.gem ffi-1.15.5.gem
Successfully installed facter-4.6.1
Building native extensions. This could take a while...
Successfully installed ffi-1.15.5
2 gems installed
```

Then facter works with 32 bit ruby.
